Realm/RLMCollection.mm

```diff
@@ -211,6 +234,17 @@ NSUInteger RLMFastEnumerate(NSFastEnumerationState *state, NSUInteger len, id<RL
                 [array addObject:list];
             }
             return array;
+        } else if (prop && prop.set && prop.swiftIvar) {
```

Comment:
The duplication here could be eliminated by renaming `initWithList:` and `initWithSet:` to something like `initWithBackingCollection:`, and then extracting this code to a function templated on the type. For just list/set this is kinda questionably worth doing, but since we know that we're adding a third container type it's worth doing some refactoring for that now.

Reply:
I've condensed this down to 
```
        if (prop && prop.collection && prop.swiftIvar) {
            get_collection_type(prop, [&](auto t) {
                for (size_t i = 0; i < count; ++i) {
                    id<RLMCollectionBase> base = [[cls alloc] init];
                    base._rlmCollection = RLMManagedCollectionFromCollection<std::decay_t<decltype(*t)>>(info, collection.get(i), prop);
                    [array addObject:base];
                }
            });
```

the managed collection is constructed with
```
template<typename Collection>
id RLMManagedCollectionFromCollection(RLMClassInfo& info, realm::Obj&& obj, RLMProperty *prop) {
    if constexpr(std::is_same_v<Collection, realm::object_store::Set>) {
        return [[RLMManagedSet alloc] initWithBackingCollection:Collection(info.realm->_realm,
                                                                           obj,
                                                                           info.tableColumn(prop))
                                                     parentInfo:&info
                                                       property:prop];
    }
    else if constexpr(std::is_same_v<Collection, realm::List>) {
        return [[RLMManagedArray alloc] initWithBackingCollection:Collection(info.realm->_realm,
                                                                             obj,
                                                                             info.tableColumn(prop))
                                                       parentInfo:&info
                                                         property:prop];
    }
    else {
        REALM_UNREACHABLE();
    }
}
```
Is there a way to not require the use of `constexpr` in this case? The compiler gives out that it can't differentiate between `realm::List` and `realm::object_store::Set` when using an ambiguous `Class`

Reply:
```
template<typename ObjcCollection, typename Collection>
id RLMManagedCollectionFromCollection(RLMClassInfo& info, realm::Obj&& obj, RLMProperty *prop) {
    return [[ObjcCollection alloc] initWithBackingCollection:Collection(info.realm->_realm,
                                                                        obj,
                                                                        info.tableColumn(prop))
                                                  parentInfo:&info
                                                    property:prop];
}
```
should work. The key is that it needs to be passed as a c++ template parameter rather than as a type-erased Class.

Reply:
nice, I've updated it